compiler/src/dotty/tools/repl/ReplCompiler.scala

```diff
@@ -88,6 +88,14 @@ class ReplCompiler extends Compiler {
     var valIdx = state.valIndex
     val defs = new mutable.ListBuffer[Tree]
 
+    def resNum(name: Name) = name.show.stripPrefix(str.REPL_RES_PREFIX).toIntOption
+
+    def maybeBumpValIdx(tree: Tree): Unit = tree match
+      case tuple: Tuple   => for t <- tuple.trees do maybeBumpValIdx(t)
+      case patDef: PatDef => for p <- patDef.pats do maybeBumpValIdx(p)
+      case tree: NameTree => for n <- resNum(tree.name) if n >= valIdx do valIdx = n + 1
```

Comment:
This is a nice one-liner but honestly I'd find a pattern match on Some/None easier to read, your call though.